libs/geometry/src/TangentSpaceMesh.cpp

```diff
@@ -187,14 +171,157 @@ Algorithm selectAlgorithm(TangentSpaceMeshInput *input) {
     return outAlgo;
 }
 
-#undef IS_INPUT_TYPE
+// The paper uses a Z-up world basis, which has been converted to Y-up here
+void frisvadKernel(const float3& N, float3& T, float3& B) noexcept {
```

Comment:
We don't have an official rule/style for output parameters like here, but I've been starting to use the return-by-value pattern. in this case it would be like this:

```
std::pair<float3, float3> kernel(float3 const& N) noexcept {
    ...
    return {T, B};
}

// calling site becomes:

auto [t, b] = kernel(n);
```

I like how it looks better. But I'll let you decide if you want to do it this way.